intro-pkgs.Rmd

```diff
@@ -21,6 +21,31 @@ What are the challenges of HTTP testing?
 * Having tests depend on having secrets for authentication at hand is not ideal.
 * Having tests for situations that are hard to trigger (e.g. the failure of a remote server) is tricky.
 
+## Tradeoffs
+
+The below packages help you meet these challenges by faking web servers or mocking their responses.
+Both approaches let you run your tests against a state of an external HTTP API which you fully control.
+This can be a good thing, because it allows you to test against all theoretically possible behavior of the API, as opposed to just the current behavior of the live API.
+
+More fundamentally, by taking control of the API behavior, your can test your code, and only your code.
+If you are testing against a live API, a failed test may be either due to your code, or the errant behavior of the API.
+This ambiguity is undesirable.
+[Unit tests](https://en.wikipedia.org/wiki/Unit_testing) should test the behavior of your code, and [integration tests](https://en.wikipedia.org/wiki/Integration_testing) should test how your code interacts with other software or services, in this case, the API [].
+^[Integration tests are sometimes further differentiated in [narrow and broad tests](https://martinfowler.com/bliki/IntegrationTest.html), where only the latter run against live services. `{vcr}` has a handy facility to switch between live and pre-recorded responses.]
+
+Unfortunately, mocking and faking add their own complexity to a project.
+Respectively, you have to carry expected responses or even (rudimentary) API behavior in your code base.
+You now have two relationships to keep in mind: your code must pass your mocked/faked tests, and your mocks/fakes must correspond to the live API.
+
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
+All other functions can then be tested without dependence on the API.
```

Comment:
Perhaps a tradeoff should be discussed for this point: I assume not testing a function against an API means testing it against some kind of test double or do you mean something else? And test doubles can at some point in the future not reflect what the API currently returns. In `vcr` this is what `re_record_interval`   https://docs.ropensci.org/vcr/reference/vcr_configure.html?q=interval#arguments attempts to achieve (TBH though i've not used this very much, mostly b/c I've not got around to setting it up)

Reply:
I've clarified the language and example (hopefully)